utils/build_stig_control.py

```diff
@@ -0,0 +1,114 @@
+#!/usr/bin/env python
+
+from __future__ import print_function
+
+import argparse
+import os
+
+import xml.etree.ElementTree as ET
+import yaml
+
+import ssg.build_yaml
+import ssg.oval
+import ssg.build_remediations
+import ssg.products
+import ssg.rules
+import ssg.yaml
+from utils.rule_dir_json import walk_products
+
+
+SSG_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
+BUILD_OUTPUT = os.path.join(SSG_ROOT, "build", "stig_control.yml")
+
+
+def parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("-r", "--root", type=str, action="store", default=SSG_ROOT,
+                        help="Path to SSG root directory (defaults to %s)" % SSG_ROOT)
+    parser.add_argument("-o", "--output", type=str, action="store", default=BUILD_OUTPUT,
+                        help="File to write json output to (defaults to build/rule_dirs.json)")
+    parser.add_argument("-p", "--product", type=str, action="store", required=True,
+                        help="What product to get STIGs for")
+    parser.add_argument("-m", "--manual", type=str, action="store", required=True,
+                        help="Path to XML XCCDF manual file to use as the source of the STIGs")
+
+    return parser.parse_args()
+
+
+def collect_rule_ids_and_dirs(rules_dir):
+    for rule_dir in sorted(ssg.rules.find_rule_dirs(rules_dir)):
+        yield ssg.rules.get_rule_dir_id(rule_dir), rule_dir
+
+
+def handle_rule_yaml(product_list, product_yamls, rule_id, rule_dir, guide_dir):
+    rule_obj = {'id': rule_id, 'dir': rule_dir, 'guide': guide_dir}
+    rule_file = ssg.rules.get_rule_dir_yaml(rule_dir)
+
+    prod_type = product_list[0]
+    product_yaml = product_yamls[prod_type]
+
+    rule_yaml = ssg.build_yaml.Rule.from_yaml(rule_file, product_yaml)
+    rule_obj['references'] = rule_yaml.references
+
+    return rule_obj
+
+
+def main():
```

Comment:
I’d suggest moving main down so it is the last method.

Reply:
I guess I should say, generally we order top->bottom where bottom is main and it is a toposorted call tree… if you call a method a from method b, a should be above b. ;-0 `get_platform_rules` and some others are mis-organized. :-)